man/olad.1
.TH olad 1 "August 2018"
.SH NAME
olad \- The Open Lighting Architecture Daemon
.SH SYNOPSIS
.B olad
[options]
.SH DESCRIPTION
.B olad
is the Open Lighting Architecture (OLA) daemon. It handles communication
with DMX512 / RDM hardware devices and remote network hosts via various
lighting control protocols.
.SH OPTIONS
.IP "-c, --config-dir <string>"
The path to the config directory, defaults to ~/.ola/ on *nix and %LOCALAPPDATA%\.ola\ on Windows.
.IP "-d, --http-data-dir <string>"
The path to the static www content.
.IP "-f, --daemon"
Fork and run as a background process.
.IP "-h, --help"
Display the help message
.IP "-i, --interface <string>"
The interface name (e.g. eth0) or IP address of the network interface to use for the web server.
.IP "-l, --log-level <int8_t>"
Set the logging level 0 .. 4. See LOGGING.
.IP "-p, --http-port <uint16_t>"
The port to run the HTTP server on. Defaults to 9090.
.IP "-r, --rpc-port <uint16_t>"
The port to listen for RPCs on. Defaults to 9010.
.IP "-v, --version"
Print
.B olad
version information
.IP "--no-http"
Disable the HTTP server.
.IP "--no-http-quit"
Disable the HTTP /quit handler.
.IP "--no-register-with-dns-sd"
Don't register the web service using DNS-SD (Bonjour).
.IP "--no-use-async-libusb"
Disable the use of the asynchronous libusb calls, revert to synchronous
.IP "--no-use-epoll"
Disable the use of epoll(), revert to select()
.IP "--no-use-kqueue"
Disable the use of kqueue(), revert to select()
.IP "--pid-location <string>"
The directory containing the PID definitions.
.IP "--scheduler-policy <policy>"
The thread scheduling policy, one of {fifo, rr}.
.IP "--scheduler-priority <priority>"
The thread priority, only used if --scheduler-policy is set.
.IP "--syslog"
Send to syslog rather than stderr.
.SH LOGGING
.B olad
can either log to
.BR stderr(4)
or
.BR syslog(5).
Each log level includes those of higher severity.
.IP 0
No logging.
.IP 1
Fatal logging.
.IP 2
Warnings.
.IP 3
Informational logging.
.IP 4
Debug logging.
.SH SIGNALS
Once running, the following signals control the behavior of olad:
.IP "SIGHUP"
Reloads the plugins.
.IP "SIGUSR1"
Increases the logging level, eventually wrapping
.SH SEE ALSO
.BR ola_dmxconsole(1) ,
.BR ola_dmxmonitor(1) ,
.


NEWS
x/y/2019 ola-0.10.8
 Features:
 * Support DragonFly BSD

 API:
 * 

 RDM Tests:
 * Fix a bug in the RDM tests to ensure they can cope with 255 personalities
   #1508
 * Only add valid network interface IDs to the test list so future tests don't
   fail
 * Add a label to the GetOutOfRangePersonalityDescription test

 Bugs:
 * Fix a few more minor documentation spelling errors found by codespell
 * Point to the main site rather than the wiki for RDM test help
 * Fix the build on FreeBSD 10.0, 10.4 and 11
 * Make PID store loading more robust
 * Clarify some of the man page options for olad
 * Fix some dead links in documentation
 * Properly close socket on destruction of Python OlaClient
 * Fix some Python 3.2 compatibility
 * Try and fetch index and type via sockaddr_dl struct for better *BSD support
 * Use C++11 if required by random #1477
 * Correct uartdmx plugin documentation for the example UART #1320
 * Ensure we don't return unsuitable network indices via RDM #1478
 * Ensure a non-existent UID exits with a non-zero exit code
 * Fix a bug in how we simulate RDM responses and collisions
 * Fix an RDM discovery bug for how we handle phantom UIDs
 * Ensure _close_callback is assigned before we might need to run it in the
   Python API

 Internal:
 *

13/7/2018 ola-0.10.7
 Features:
 * Allow multiple KiNet ports to be patched to the same universe #1414
 * Allow multiple FTDI ports to be patched to the same universe
 * Support libftdi1 as well as libftdi0 in FTDI DMX #1012 Debian #810374
 * Support FTDI DMX via libftdi1 on Windows

 RDM Tests:
 * Fix a bug in the RDM tests when a responder supports sensor PIDs #1388
 * Fix a bug in imports for rdm_responder_test.py

 Bugs:
 * Fix the build on Windows
 * Entire codebase now passes codespell testing
 * Fix a Python 3 bug in the API
 * Fix a few longstanding bugs in the Python ola_rdm_discover
 * Correct the OSC config option name for output format in the documentation
 * Use top_builddir in more places in the build
 * Use libusb_set_option where available
 * Fix the udev rules
 * Downstream Debian fixes
 * Fix an incorrect response if we have the sensor PIDs but no sensors
 * Stop trying to build opendmx kernel plugin on Windows
 * Prefer Avahi over Bonjour for our DiscoveryAgent, in case the Bonjour
   version is actually just Avahi's compatibility layer
 * Clean up the libolaserver pkg-config file

 Internal:
 * Various minor log formatting tidying up

07/01/2018 ola-0.10.6
 Bugs:
 * Renamed fempto(sic) to femto in the RDM code (C++ and Python)
 * Renamed terra(sic) to tera in the RDM code (C++ and Python)
 * Fixes for RDM test server forward compatibility with later JQuery versions
 * Fix a man page typo
 * Fix a longstanding copy-paste typo in the Python client's RegisterUniverse
   function
 * Increment the Transaction Number for each Art-Net RDM packet #189
 * Fix lots of minor documentation spelling errors found by spellintian and
   codespell
 * Renamed protocol convertor(sic) to protocol converter in the RDM code (C++
   and Python)

17/07/2017 ola-0.10.5
 Features:
 * Make some error logs more verbose

 Bugs:
 * C++11 compatibility fix (for cppunit 1.14.0) #1244
 * Always print frames in ola_rdm_get/set, even if we get an error
 * Ensure man files ship in the correct packages #1276
 * Check protobuf is old enough, until we fix #1192

21/05/2017 ola-0.10.4
 Features:
 * Add more help regarding the ArtNet use_limited_broadcast option

 Bugs:
 * Ensure Pathport node actually uses configured IP #1166
 * Improve some debugging around flag parsing #1198
 * Disable warnings for readdir_r deprecation until we can rewrite our code to
   stop using it. Related to #1055
 * Stop replacing -I with -isystem and correctly work around the resultant
   errors #1125
 * Exit, rather than continuing, if ola_throughput failed to send DMX
 * Implement OLA_FALLTHROUGH macro for adding fallthrough attributes to 
   switchcases in gcc versions higher than 7 #1193 Debian #853583
 * Fix Protobuf 2.6 not building on the latest OS X
 * Update some URLs in the Windows build documentation
 * Skip interface count checks on Windows, as it may fail them if it's not on a
   network due to a lack of loopback interface
 * Fix LibUSB code on Windows, especially when the Hotplug functionality exists
   but isn't available
 * Support wireless interfaces on Windows since Vista
 * Add a cast to fix an error when running execlp
 * Ensure configure checks for tcmalloc's presence before adding it to the libs
   when asked for #1201
 * Add some more detail to the Logic RDM sniffer and Enttec RDM USB Pro sniffer
   documentation

06/12/2016 ola-0.10.3
 Features:
 * Add a message explaining why the auto start test may fail

 Bugs:
 * Work around a clang warning in our protobuf generator code
 * Force gnu++98 mode if compiler supports 98 and 11 modes so we keep
   std::auto_ptr #1028 Debian #831103
 * JsonPatchParser.h tries to include a project header file; move the
   OptionalItem header to includes to resolve #1086
 * Reset the client socket once we've closed it, so it can be re-used
 * Silence a warning in the code so it builds with Protobuf 3 #1110 Debian
   #835433
 * Work around <ernno.h> include failure on MIPS Debian #836383
 * Update GPIO and SPI pin limits for CHIP as its pin numbers are higher #1095
 * Clarify some details in ola_set_dmx and ola_streaming_client help #1109
   Debian #830972
 * Fix a logging error regarding USB product name detection

 Internal:
 * Add more debugging of opening and acquiring locks

21/5/2016 ola-0.10.2
 Features:
 * Add support for Enttec USB Pro Mk 2 B
 * Make DNS-SD functionality optional at build time
 * Name multi-port Enttec Pro Mk 2 ports to match the labelling on the device

 RDM Tests:
 * Correct expected NAck type in GetSettingDescriptionsMixin, as spotted by ETC
 * Fix the mute all devices test for controllers that don't support discovery

 Bugs:
 * Allow one slot devices to correctly be set to any start address, including

olad/Olad.cpp
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Olad.cpp
 * Main file for olad, parses the options, forks if required and runs the
 * daemon.
 * Copyright (C) 2005 Simon Newton
 *
 */

#if HAVE_CONFIG_H
#include <config.h>
#endif  // HAVE_CONFIG_H

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include <iostream>
#include <memory>

// On MinGW, OlaDaemon.h pulls in SocketAddress.h which pulls in WinSock2.h,
// which needs to be after WinSock2.h, hence this order
#include "olad/OlaDaemon.h"

#if HAVE_LIBSYSTEMD
#include "olad/Systemd.h"
#endif  // HAVE_LIBSYSTEMD

#include "ola/Logging.h"
#include "ola/base/Credentials.h"
#include "ola/base/Flags.h"
#include "ola/base/Init.h"
#include "ola/base/SysExits.h"
#include "ola/base/Version.h"
#include "ola/thread/SignalThread.h"

using ola::OlaDaemon;
using ola::thread::SignalThread;
using std::cout;
using std::endl;

DEFINE_default_bool(http, true, "Disable the HTTP server.");
DEFINE_default_bool(http_quit, true, "Disable the HTTP /quit handler.");
#ifndef _WIN32
DEFINE_s_default_bool(daemon, f, false,
                      "Fork and run as a background process.");
#endif  // _WIN32
DEFINE_s_string(http_data_dir, d, "", "The path to the static www content.");
DEFINE_s_string(interface, i, "",
                "The interface name (e.g. eth0) or IP address of the network "
                "interface to use for the web server.");
DEFINE_string(pid_location, "",
              "The directory containing the PID definitions.");
DEFINE_s_uint16(http_port, p, ola::OlaServer::DEFAULT_HTTP_PORT,
                "The port to run the HTTP server on. Defaults to 9090.");

/**
 * This is called by the SelectServer loop to start up the SignalThread. If the
 * thread fails to start, we terminate the SelectServer
 */
void StartSignalThread(ola::io::SelectServer *ss,
                       SignalThread *signal_thread) {
  if (!signal_thread->Start()) {
    ss->Terminate();
  }
}


/*
 * Main
 */
int main(int argc, char *argv[]) {
  // Take a copy of the arguments otherwise the export map is incorrect.
  const int original_argc = argc;
  char *original_argv[original_argc];
  for (int i = 0; i < original_argc; i++) {
    original_argv[i] = argv[i];
  }

  // We don't use the longer form for ServerInit here because we need to check
  // for root and possibly daemonise before doing the rest of the work from
  // ServerInit.

  ola::SetHelpString("[options]", "Start the OLA Daemon.");
  ola::ParseFlags(&argc, argv);

  ola::InitLoggingFromFlags();
  OLA_INFO << "OLA Daemon version " << ola::base::Version::GetVersion();

  #ifndef OLAD_SKIP_ROOT_CHECK
  uid_t uid;
  ola::GetEUID(&uid);
  if (ola::SupportsUIDs() && !uid) {
    OLA_FATAL << "Attempting to run as root, aborting.";
    return ola::EXIT_UNAVAILABLE;
  }
  #endif  // OLAD_SKIP_ROOT_CHECK

#ifndef _WIN32
  if (FLAGS_daemon)
    ola::Daemonise();
#endif  // _WIN32

  ola::ExportMap export_map;
  if (!ola::ServerInit(original_argc, original_argv, &export_map)) {
    return ola::EXIT_UNAVAILABLE;
  }

  // We need to block signals before we start any threads.
  // Signal setup is complex. First of all we need to install NULL handlers to
  // the signals are blocked before we start *any* threads. It's safest if we
  // do this before creating the OlaDaemon.
  SignalThread signal_thread;
  signal_thread.InstallSignalHandler(SIGINT, NULL);
  signal_thread.InstallSignalHandler(SIGTERM, NULL);
#ifndef _WIN32
  signal_thread.InstallSignalHandler(SIGHUP, NULL);
  signal_thread.InstallSignalHandler(
      SIGUSR1, ola::NewCallback(&ola::IncrementLogLevel));
#endif  // _WIN32

  ola::OlaServer::Options options;
  options.http_enable = FLAGS_http;
  options.http_enable_quit = FLAGS_http_quit;
  options.http_port = FLAGS_http_port;
  options.http_data_dir = FLAGS_http_data_dir.str();
  options.network_interface = FLAGS_interface.str();
  options.pid_data_dir = FLAGS_pid_location.str();

  std::auto_ptr<OlaDaemon> olad(new OlaDaemon(options, &export_map));
  if (!olad.get()) {
    return ola::EXIT_UNAVAILABLE;
  }

  // Now that the OlaDaemon has been created, we can reset the signal handlers
  // to do what we actually want them to.
  signal_thread.InstallSignalHandler(
      SIGINT,
      ola::NewCallback(olad->GetSelectServer(),
                       &ola::io::SelectServer::Terminate));
  signal_thread.InstallSignalHandler(
      SIGTERM,
      ola::NewCallback(olad->GetSelectServer(),
                       &ola::io::SelectServer::Terminate));

  // We can't start the signal thread here, otherwise there is a race
  // condition if a signal arrives before we enter the SelectServer Run()
  // method. Instead we schedule it to start from the SelectServer loop.
  olad->GetSelectServer()->Execute(ola::NewSingleCallback(
        &StartSignalThread,
        olad->GetSelectServer(),
        &signal_thread));

  if (!olad->Init()) {
    return ola::EXIT_UNAVAILABLE;
  }

#ifndef _WIN32
  // Finally the OlaServer is not-null.
  signal_thread.InstallSignalHandler(
      SIGHUP,
      ola::NewCallback(olad->GetOlaServer(), &ola::OlaServer::ReloadPlugins));
#endif  // _WIN32

#if HAVE_LIBSYSTEMD
  if (ola::SystemdNotifyAvailable()) {
    OLA_INFO << "Systemd notification socket address present, "
             << "sending notifications.";
  } else {
    OLA_WARN << "Systemd notification socket address not present, "
             << "not sending notifications.";
  }
  // Does not need to be guarded. sd_notify does its own internal check on
  // the socket's presence, as well.
  ola::SystemdNotify(0, "READY=1\nSTATUS=Startup complete\n");
#endif  // HAVE_LIBSYSTEMD
  olad->Run();
  return ola::EXIT_OK;
}
